### DISCUSSION TITLE
The future of ESPAltherma

### DISCUSSION THREAD

**Discussion by raomin** [raomin/ESPAltherma#252] [Polls]
Hello everyone!

I've been considering several options for developing further ESPAltherma and I'd like your opinion.
I'll briefly describe each of them and I'll let you vote for the one you prefer. I'm not saying I'll follow immediately with implementation but your votes will definitely guide me (and possible other contributors!).

### Heat pump Autodetect
Seeing the numerous _issues_ with users unsure on which definition file should be used, I believe we should simplify this...
There are very minor differences between the definitions, mainly 3-4 big families. I guess we could design a default _auto-detect_ mode that would identify which protocol (I/S) and which HP family (High temp, geo, gas, low temp) from probing specific registers.
This would require some support from some of you but I guess we can figure this out and that would greatly help.

### Web installer
Still trying to facilitate adoption here. The current installation process might be a bit daunting with the need for Platformio, changing config files etc. even if we have numerous positive feedbacks from First-time-hackers.
ESPAltherma could integrate a web installer that would deploy from the browser the firmware to the esp device, using [esp-web-tools](https://esphome.github.io/esp-web-tools/). This would be facilitated with the Autodetect feature so that there would be only a couple of firmware to maintain.

### ESPHome integration
ESPAltherma as an espHome component. Not sure yet how it would work in practice but I assume this could be done.

### A custom ESPAltherma board
Why not. A crstom board with an ESP32, the right connectors (!), and a set of relays to connect as thermostat, to the smart grid SG1/SG2... Sold at a reasonable price, maybe crowdfunded?

### Fine Consumption control through X85A + Pulse meter
There are ways that ESPAltherma could control the power consumption of the HP. Through [the X85A+Pulse meter](https://github.com/raomin/ESPAltherma/discussions/170) with simple GPIOs and or through the SG1/SG2 port (already supported). ESPAltherma acting as a pulse meter of solar installation would allow for fine precision in the power consumption. 

### HomeAssistant improved integration
Declaring manually each sensor in HA is not really a good user experience and could be avoided with a better integration with HA. Either through an integration plugin or through the automated declaration of all values as sensor in MQTT. That would make a lot of sensors though.

### Other ideas?
Add your own in comments below and let's discuss!

**Comment by colseven** [comment]
Control of the hot water temperature in the domestic hot water storage tank.

**Reply by GO74** [reply]
That would be really nice.

**Reply by GO74** [reply]
We could trick the System if we would switch between a resistor based voltage divider (simulating >60°C) and the actual probe. Altherma wouldn't heat up the storage if the (fake) measurement is above the desired temperature. So if we want hot water, we switch to the actual sensor. We could then either heat it up to 60°C to disinfect the system (every couple of weeks), or switch back to the fake measurement when we reached the desired temperature.

**Reply by javimurcia** [reply]
I think that could be more easily accomplished using the P1P2 interface (that allows you to change your DHW setpoint at will, if your heat pump is supported) or via the cloud (i think there are libraries that allows you to interact with daikin's cloud to change your DHW setpoint)

Also, it depends on what you want to archive, if you want to produce hot water on demand, you can configure your heat pump with 2 setpotins for DHW (an eco setpoint and a confort setpoint).
You can use the smart grid interface to make the heat pump switch to the confort setpoint.

**Reply by GO74** [reply]
Well, mine isn't connected to any cloud and as far as I know ESPAltherma can (only) read whatever is on the x10a. So the easy option for me would be switching back an forth to a fake temp sensor.

**Comment by GO74** [comment]
How about sending COP and doing the calculations directly on the ESP? Small feature, but why not?

**Comment by alexdelli** [comment]
I think the ESPhome integration is the future. I have a lot of esp32 rpeater in my house, so signal is everywhere without using wifi repeater.
The integrations could be made via WEbGUI: this is an example:
https://esphome.github.io/bluetooth-proxies/

**Comment by ap20132022** [comment]
how about integration + coorperation with pyHSPU?
Actually I need pyHSPU for some more informations (valve, water pressure...). I need to install both devices to get all the information. It will be wonderfull if only one device needed. 
ESP doesn't have CAN-Bus. So I think we need here raspberry or a pc.

**Reply by KalaNagHTD** [reply]
I also use pyHPSU in addition. The downside of such an integration is that just the HPSU has the CAN-Bus. So the pyHPSU integration has no benefit for the majority of Altherma owners.

**Reply by javimurcia** [reply]
The esp32 has a can controller, it's only missing the Transceiver, so i don't see why this cannot be implemented on a esp32.

BTW, check this https://github.com/Arnold-n/P1P2Serial maybe it can querry the information you are seeking

**Reply by KalaNagHTD** [reply]
> BTW, check this https://github.com/Arnold-n/P1P2Serial maybe it can querry the information you are seeking

This serial bus is for the Daikin air conditioners and not implemented with the HPSU/ECH2O built before 2/2022. Even for the recent version I‘m not sure.

**Comment by amasciotta** [comment]
Since I had (and still have) a lot of trouble to choose the right setup file, I think the heat pump autodetect feature would be a great addition, especially for newcomers

**Comment by Ripley42** [comment]
My installer failed to install an external thermistor even though the outdoor unit catches direct sunlight and reports wrong temperatures. This obstructs the use of WD curves and automatic switching between heating and cooling modes. I could try to run an external thermistor to my roof myself, but I’ve been thinking about emulating the thermistor with a DAC and controlling the input with the temperature reported by a weather service. Maybe a useful addition to your project.

**Reply by javimurcia** [reply]
You don't need to place the external thermistor at the roof, you just need to place it on the north face of your house, and away from any heat source (a vent for instance), so at long its more than 1.5 meters from the ground, you are good to go.

BTW, if you want to emulate a thermistor, what you need is a digital potentiometer, you can find the temperature - resistance chart of the original thermistor at page 262 here: https://www.manualslib.com/manual/2560305/Daikin-Altherma-3-M.html?page=262

**Reply by Ripley42** [reply]
The neighbors have the North side of the house. I have a space on my roof that is completely sheltered from the sun, but it’s hard to get to with a wire. That’s why I thought of emulating the sensor. Thanks for the spec and the headsup about the potentiometer.

**Reply by abmantis** [reply]
@javimurcia  have you tried it? I was wondering about using a digital potentiometer as a fake indoor temperature sensor to use modulation: https://github.com/raomin/ESPAltherma/discussions/412

**Comment by chillymattster** [comment]
Hi there,
I don't know if this has already been discussed somewhere or if I'm to late as I found this project just a few days ago.
What about adding CAN support as well?
There is an older pull request stating that some preparations have been made. And there is a nice this nice ["ATOM CANBus Kit (CA-IS3050G)"](https://shop.m5stack.com/products/atom-canbus-kit-ca-is3050g) from M5Stack for the Atom lite which would allow the use of CAN and in addition the serial X10 connection via Grove cable.
In my opinion this would be a great chance to close the gap and finally get all the values like water pressure and precise valve positions out of our devices as well as opening the opportunity to start controlling our heat pumps via CAN, e.g. setting operating modes and changing temperatures.

**Comment by munzili** [comment]
Just to add my peanuts

In a couple of days will be my PR #110 ready for a "possible" merge. ATM i'm merging changes from the main branch into my work. Some minor tests also still needs to be done and small changes are open

This will contain the webui features, CAN Feature and other stuff but also will/may create backwards incompatibility (Not sure if ESP8266 still can be somehow supported, i'm still looking into this).

**Reply by maromme** [reply]
that sounds great. would you mind implementing the i2c library for the port expander as I added in #273 (it's for sure in a nicer way possible than done by me). For those using the m5stick the three "free" GPIOs might not be enough. I at least neet SG1/SG2/heat/cool and the x85a feature #262 Means 8 GPIOs. I merged the for my use the main branche #273 and #262 into a running system. Just the colling feature missing

**Reply by munzili** [reply]
All open PR #260, #262, #273 are easy to be merged.

**Reply by maromme** [reply]
if you need someone testing I'd volunteer

**Comment by Kepro** [comment]
I would say ESPHome will simplify things and also integration with HA

**Reply by neildsb** [reply]
+1 ESPHome Please :-)

**Comment by mdr-github** [comment]
This is an old thread but I'm new to the project, just installed my ESPAltherma yesterday :)

The web installer would be nice. I was able to manage with PlatformIO but only after much fiddling. There is a learning curve since this was my first project, but I worked it out. I couln't get the virtual COM port to work on Windows 11 and had to resort back to another computer running Windows 10. These issues aren't project specific, but still a web installer would be nice for newcomers.

I see a lot of votes for autodetect, but think this 'issue' can also be resolved with just a bit more documentation. I'm assuming you know what model heat pump you have (which you should if you've spent that much cash on one). Maybe I'm oversimplifying, but then it's just a matter of looking up the model using the Daikin naming scheme.  But on that topic, a bit more documentation on the various MQTT values would be nice as well.

+1 on the COP value, this is a key metric and so much easier if the ESP32 can calculate and report this directly.

Keep up the good work @raomin, thanks!

**Comment by scapman** [comment]
+1 ESPHome. That would be great!
Thanks!

**Reply by fernwerker** [reply]
@raomin, is there already any esphome activity planned or ongoing?

**Comment by D-hofer** [comment]
A API interface with JSON data would be great. Then you can easily integrate the values to other home automation system directly without an additional server/bridge/...

**Comment by fernwerker** [comment]
By the way, I think the votes for "HomeAssistant improved integration" and "ESPHome integration" can be consolidated as ESPHome brings a better HomeAssistant integration.

So, this joint approach would have >60% votes. 😄

**Comment by engelchrisi** [comment]
You can try my esphome integration based on the ESPAltherma code at your own risk of course. You need some experience with custom components in esphome to do so:

https://github.com/engelchrisi/ESPAltherma

**Reply by Zipponia** [reply]
do you have an example code we can use with Esphome ?

**Comment by KalaNagHTD** [comment]
The ESPHome solution for Daikin HPSU/ECH2O units does exist already:
[https://github.com/Trunks1982/Daikin-Rotex-HPSU-CAN-Seriell](url)

**Comment by Crucio32000** [comment]
Hi everyone.
I am using this project for a year and its working flawlessly.
So far so good, i believe another possible step forward would be the ability to change some data, not only monitor.
Perhaps we could reuse the reverse engineering done for other projects as well (See p1p2Serial). It seems that the protocol layout is similar.
Is this something appealing to you guys ?
I believe the X10A should be able to also set some data.

**Reply by KalaNagHTD** [reply]
Unfortunately the X10A is not made for setting data!

**Reply by Crucio32000** [reply]
Ah! Do we have any evidence about this ?
Cause the protocol layer seems to be very similar to the P1P2.
If there is a slight chance that we can tune some fields, perhaps its worth investigating this further. 
We could start adding a topic over MQTT to test some packet sending by indicating the payload to be sent and expected timeout for reading the response. 
Of course a flush prior and after so we are sure not to messing around with other working queries.

**Reply by KalaNagHTD** [reply]
For me the evidence is that Daikin has been using this X10A service interface for nearly 20 years with all his AC- and heating devices while countless people have tried controlling their units with it. Even the Daikin service software can not control the units with the X10A interface. It can just load complete parameter sets after having put the mainboard in flash mode.

**Reply by Crucio32000** [reply]
It makes sense.
The thing is that it is not explicitly told anywhere that attempts were made, without any result.
So perhaps the fact that this X10A interface has been available since forever, and nobody was able to do so, provides a good confidence that control is not viable with this bus.

How are you guys achieving control ? In general, i would like to to be able to change the DWT / LWT setpoint.

I was considering to buy , for instance, a BRP069A81 WiFi module.

Whats your take on this regard ?

It could also be integrated in the README, indicating that control can be achieved by other means.

**Reply by Crucio32000** [reply]
Checking within Daikin product, it seems that X10A is used to connect the heatpump to a PC.
See here [EKPCCAB4](https://www.daikin.it/content/dam/document-library/installation-manuals/ctrl/EKPCCAB4_4PW74106-1E_2018_06_Installation%20manual.pdf)

**Reply by pshotton** [reply]
> 
> How are you guys achieving control ? In general, i would like to to be able to change the DWT / LWT setpoint.
> 
> I was considering to buy , for instance, a BRP069A81 WiFi module.

I've got the BRP069A82 wifi module (don't know how it differs from the A81 version). It allows remote control from the Daikin app, setting heat and turning unit off/on. Once you have one of these you can make websocket calls to both read and control your unit.
There's plenty of code out there, see for instance https://github.com/Frankkkkk/python-daikin-altherma

**Comment by divenal** [comment]
Hi, I've had HP for just over a year, and have meaning to deploy this (and definitely should have done it over the summer!).
A few thoughts...

1. Rather than developing against a live system, would be worth knocking up a trivial (dumb) emulator implementing the protocol. Eg using an ftdi usb serial adapter can provide the same pins as the X10A, and then a trivial python script can respond correctly to the protocol. (And even allows testing of dropped characters, incorrect checksum, etc.) If anyone can spare raw serial output from live systems, that would be helpful.
2. I think it might be useful to add support for a humidity sensor (and maybe temperature) - this already gives us a powered esp32 outside the house, so adding support for a sensor saves having to set up a second external device.
3. But that gets me wondering whether to use the original code base, or an ESPHome-based alternative.

I have to say I was rather disappointed to the see the hard-coded converter id values in the tables. I'd have thought symbolic entries like CONV_LE16_SCALE256  would make more sense.  Also disappointed by the duplication - lots of duplicate code to read big- or -little-endian, and then scale by 2, 10, 256, etc.  Could have coded that as bit-fields within an integer, so 1 bit gives endianness, and maybe three bits to encode the scaling factor, that sort of thing. Then converter could be data-driven by these bits. Would be a lot of work (and high risk) to change it in ine step, though.  But could first convert everything to symbolic entries using the existing converter values, and then once that's complete, would be easier to drop in a replacement encoding.

I don't use Home Assistant, but quite impressed by esphome facilities and ease of use / deployment. It does all the work of setting up build environment, etc. So that might be a solution to the complexity mentioned in the opening post. (I just use esphome with the python client at the host end.)

(I had been looking for a way to avoid setting up an mqtt server just for this. Easy enough, and I already have a 24/7 rpi doing stuff, so that's not an issue. Just seems a bit heavyweight (which is ironic, since mqtt is a protocol designed for IOT).  mqtt-sn would perhaps be more appropriate, but doesn't really seem to have caught on.  I had actually been contemplating just modifying this code to send the raw serial packets out as udp multicasts, and then have my client code receive and decode them. In fact raw ethernet frames would be better, but esp libraries don't seem to offer sending raw frames over wifi. Not sure why...  But I digress - I don't suppose the mqtt aspect is an issue for most users.)

**Comment by divenal** [comment]
Just for fun, I've been having a play with ESPHome to see if I can get the serial interface working. I have something that appears to work, though note that I've not tested it on a live system, just with a python script at the other end of the serial sending back canned responses. It has got to the point where I can deliver the output from the serial into a text sensor (hex-encoded). I think that's actually enough for my needs - I can do the decoding and conversion on the host side, rather than on the esp32.

UPDATED: found a _much_ tidier solution:

```
# define the uart connection to the X10A serial, and use the
# debug mechanism to get hold of the received data. Each time
# we receive a message, hex-encode it and send it to the text sensor,
# and request the next registry in sequence.

uart:
  tx_pin: GPIO1
  rx_pin: GPIO3
  baud_rate: 9600
  data_bits: 8
  parity: EVEN
  stop_bits: 1
  id: x10a
  debug:
    dummy_receiver: true
    direction: RX
    after:
      timeout: 100ms
    sequence:
      # TODO: validate the response
      - lambda: id(rawaltherma).publish_state(esphome::format_hex(bytes));
      - script.execute: reader

globals:
    # This records the index into the table of the next registry to request
  - id: regidx
    type: int
    restore_value: no
    initial_value: '-1'

    # This is the register currently being requested
  - id: reg
    type: int
    restore_value: no
    initial_value: '-1'

# This tries up to 3 times to retrieve one register.
# Starting, or restarting, the script is the mechanism to
# move onto the next one.
# It doesn't itself know when the response arrives,
# but the uart component simply restarts the script
# when it does arrive.

script:
  - id: reader
    mode: restart
    then:
      # Choose next registry to read
      - lambda: |-
         static const uint8_t registers[] = { 0x60, 0x61, 0x10, 0x62 };
         unsigned int idx = ++id(regidx);
         if (idx >= sizeof(registers)) {
            ESP_LOGI("reader", "done");
            id(reader)->stop();
         } else {
            id(reg) = registers[idx];
         }

      # Request the data up to 3 times, with 2 second delay between retries
      # If the data arrives, the script will be restarted, which moves us
      # onto the next register.

      - repeat:
          count: 3
          then:
            - uart.write: !lambda
                unsigned char x = (unsigned char)id(reg);
                ESP_LOGD("reader", "requesting 0x%x", x);
                return {0x03, 0x40, x, (unsigned char)(0xff - 0x03 - 0x40 - x)};
            - delay: 2s
            
      - logger.log: "No response from x10a"

      # Restart the script to move onto the next register
      - script.execute: reader

# we hex-dump data received from the uart into this sensor.
# (Is there an equivalent that would just let us use binary data ?)
# This gives a mechanism to pass the raw data over to an api client
# for decoding.

text_sensor:
  - platform: template
    id: rawaltherma
    name: "rawaltherma"
    on_value:
      then:
         - lambda: ESP_LOGD("rawaltherma", "%s", x.c_str());

# every 30 sec, we reset regidx and start the script

interval:
  - interval: 30sec
    then:
      - logger.log: "starting reader..."
      - globals.set:
          id: regidx
          value: '-1'   # it will pre-increment to 0
      - script.execute: reader

```

It should be possible to do the decoding / conversion on the esp32, and populate sensors that way. Best way is probably to include a class to contain the code, since it would be rather unweildy doing it all as inline lambda code.

Currently no checking on the received data, but that should be fairly trivial.

(But I think I ought to hold off opening up the heatpump to attach a cable for a few days - Chrismas would be a bad time to break the heating !)

**Comment by jjohnsen** [comment]
**ESPHome Altherma Now Available! 🎉**

I've had some free time so I've done a rewrite of a previous ESPHome implementation I've been running since 2022🙈.
It's now functional with:

✅ Native ESPHome custom component
✅ Web-based installer (no coding required)
✅ OTA updates support
✅ Home Assistant auto-discovery
✅ Model-specific sensor configuration support
✅ Mock UART for development

Currently supports ERGA-D EHV/EHB/EHVZ DA series (04-08kW). 
Contributions for other models welcome!

Please have a look at: https://github.com/jjohnsen/esphome-altherma

Installer at: https://jjohnsen.github.io/esphome-altherma/

**Comment by Dennis3321** [comment]
Here's my project;

https://github.com/Dennis3321/DaikinAlthermaX10AConnect

Features

- Build on ESP Home, easier to build, easier to update
- Build for a M5 stack POE (power over ethernet) device, because ‘a cable is stable’
- Ability to set the mode; off,heat, cool (if your heat pump can also cool)
- Ability to select the smart grid mode; Free running, forced off, recommend on, forced on.
- Debug mode for trouble shooting.
- Build for the m5 stack 4 channel relay, because standardized and state restore
